### PR TITLE
Autotune complete cached-attention sequences

### PR DESCRIPTION
**Pull request by kvark**
Stacked on #192, which uses ordinary command recording. No replay API or platform-specific winner.

Extend the existing bounded search to cached block attention:

- Compare 1/2/4/8/16 splits (one means the single-dispatch kernel), retaining the incumbent when qualification, timing, memory or deadline checks do not establish a gain.
- Measure the entire split/combine sequence at short/middle/full cache positions. Use private matching-placement scratch, nonzero and tiny-f32 inputs, full finite/cross-variant checks, sampled f64 attention references, and empty-partial checks after long-to-short reuse.
- Reuse staging, paired timing, phase accounting, decision/noise guards and the existing kernels. Preserve live cache/output contents; retain partial-buffer identities across repeated tuning.
- Keep attention's external bindings live across the whole sequence in memory planning. Installing a selection preserves original barrier boundaries; partial growth copies old allocation contents before replacing storage.
- Keep matrix report serialization compatible. Add `TuneScope::Attention` and `attention_outcomes`; the ordinary `All` search visits attention within the same time/class budget, after matrix classes.

### Measured results

SmolLM2-135M, one sequence, 128-token prefill and batch-one cached decode, full logits copied to the host. Three fresh processes per arm and GPU, rotated order; ordinary command recording throughout. These are diagnostic whole-call latencies, not the frozen Inferena cohort.

| GPU / phase | Before (ms) | New lifetime plan, dense search only (ms) | Full search (ms) | llama.cpp (ms) |
| --- | ---: | ---: | ---: | ---: |
| RTX 5070 prefill | 8.794 | 8.816 | 8.758 | 7.087 |
| RTX 5070 decode/token | 3.131 | 3.130 | 2.973 | 1.376 |
| Arc B570 prefill | 22.825 | 22.752 | 21.287 | 15.649 |
| Arc B570 decode/token | 5.595 | 5.612 | 5.506 | 3.438 |

Every run selected 16 NVIDIA decode splits and unsplit Intel prefill. The other two cases retained eight splits. Intel prefill improves about 7% in all three trials. NVIDIA decode improves about 5% by the median, **but one full-search process regressed** (2.970–3.239 ms versus 3.127–3.137 ms before); clocks were not fixed. Intel decode did not change attention choice, so its small whole-call difference is not a selection gain. The gap to llama.cpp remains.

Attention comparisons cost about 0.46 seconds across the two NVIDIA sessions and 1.40 seconds on Intel. Neither session exhausted its 30-second budget. All 24 saved logit sets are finite and agree with the CPU f32 reference on all 33 next-token choices. Full-search maximum per-row relative L2 is 0.00000994 / 0.000177.

Exact measured implementation: `861d87b9a85f666a5a131f82d07c0b66d9131e16`; control: `52027a8fc9e243a50039967b4d0dee0c51a05897`. The follow-up only corrects selected-kernel labels and records these results. Reproduction details and limits: [bench/cached-attention-tuning.md](https://github.com/kvark/meganeura/blob/050f8a35f33a188877321b3fbfa39d4b1194f83c/bench/cached-attention-tuning.md).

### Verification

453 CPU tests passed; the expanded existing attention regression passed on both GPUs with unchanged tolerance. The opt-in transition test forces repeated 1/2/4/16 choices and verifies output preservation and stable allocation footprint on both GPUs. Formatting, all-feature/all-target Clippy and rustdoc with warnings denied passed. All six jobs passed for the measured implementation in [CI run 737](https://github.com/kvark/meganeura/actions/runs/35467384352). [Run 738](https://github.com/kvark/meganeura/actions/runs/35467897156) also passed all six jobs for the final head. The measured revision's retained coverage artifact reports 82.8% Rust host line coverage overall and 75.6% in the new attention module. This does not instrument WGSL; the forced storage-transition test is opt-in and was run locally on both GPUs.

The known Naga Workgroup ArrayStride validation warning remains unchanged. No paper/cohort change, raw data, binary artifacts, new test executable, precision relaxation or platform threshold.

Rebased after #192 merged into main. The file tree is unchanged from the pre-rebase head; the measured revisions remain on `experiment/llama-gemv-2026-09-19`. Fresh post-rebase CI is running.

### CODE REVIEW

File: src/runtime/tuning/attention.rs [hunk 1]
```diff
@@ -0,0 +1,964 @@
+use super::*;
```
Comment (kvark) [inline src/runtime/tuning/attention.rs:1]: we should avoid wildcard dependencies in this repo. Just use `super::Xxx` to address these types

File: src/runtime/tuning/attention.rs [hunk 2]
```diff
@@ -0,0 +1,964 @@
...
+        let &[
+            window_size,
+            num_heads,
+            num_kv_heads,
+            head_dim,
+            block_len,
+            max_seq,
+            _,
+            _,
+        ] = d.params.as_slice()
```
Comment (kvark) [inline src/runtime/tuning/attention.rs:29]: wow that looks dangerous. What if the slice order changes? you'd have a subtle mistake in this code?

File: src/runtime/tuning/attention.rs [hunk 3]
```diff
@@ -0,0 +1,964 @@
...
+        else {
+            return None;
+        };
+        if [num_heads, num_kv_heads, head_dim, block_len, max_seq].contains(&0)
+            || !num_heads.is_multiple_of(num_kv_heads)
+            || head_dim > 512
+            || block_len > max_seq
+            || num_heads > 65_535
```
Comment (kvark) [inline src/runtime/tuning/attention.rs:37]: what's up with these constants? we should specify them in hex

File: src/runtime/tuning/attention.rs [hunk 4]
```diff
@@ -0,0 +1,964 @@
...
+            outcome.decision = if start.elapsed() >= options.max_time {
+                TuneDecision::TimeBudget
+            } else {
+                TuneDecision::InvalidOutput
+            };
+            outcome.failure = Some(error);
+            return;
+        }
+        outcome.qualified = true;
+        drop(qualification);
```
Comment (kvark) [inline src/runtime/tuning/attention.rs:611]: can we avoid manual drops?

File: tests/gemma_inference_ops.rs [hunk 5]
```diff
@@ -349,5 +351,37 @@ fn cached_block_writes_only_valid_rows_and_selects_last() {
             session.wait();
             assert_close(&session.read_output(2 * dim), &expected, 1e-5);
         }
+        if dim == 64 {
```
Comment (kvark) [inline tests/gemma_inference_ops.rs:354]: why 64 specifically?

**Review by kvark — COMMENTED** [review]
I feel that this implementation is overly complex. Can we simplify?

**Comment by kvark** [comment]
Superseded by #200, a single PR on current main with six focused implementation commits and one documentation commit. It includes #194/#195 (which merged into this branch), addresses the inline review on typed attention parameters, wildcard imports, lifetimes, hex limits and the representative test case, and preserves the old source/reports at experiment/llama-catchup-pre-cleanup-2026-09-20. Final combined verification is running there. Closing this review thread/PR in favor of the consolidated diff; no branch is deleted.
